compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/generators/DataClassMembersGenerator.kt

```diff
@@ -67,37 +71,83 @@ class DataClassMembersGenerator(val components: Fir2IrComponents) {
                 valueParameterDescriptor.bind(this)
             }
 
+        private val equalsName = Name.identifier("equals")
+        private val hashCodeName = Name.identifier("hashCode")
+        private val toStringName = Name.identifier("toString")
+
+        private fun checkEqualsSignature(declaration: FirSimpleFunction): Boolean =
+            declaration.valueParameters.size == 1 &&
+                    declaration.valueParameters[0].returnTypeRef.toIrType(components.typeConverter) == components.irBuiltIns.anyNType &&
+                    declaration.returnTypeRef.toIrType(components.typeConverter) == components.irBuiltIns.booleanType
+
+        private fun checkHashCodeSignature(declaration: FirSimpleFunction): Boolean =
+            declaration.valueParameters.isEmpty() &&
+                    declaration.returnTypeRef.toIrType(components.typeConverter) == components.irBuiltIns.intType
+
+        private fun checkToStringSignature(declaration: FirSimpleFunction): Boolean =
+            declaration.valueParameters.isEmpty() &&
+                    declaration.returnTypeRef.toIrType(components.typeConverter) == components.irBuiltIns.stringType
+
         fun generate(): List<Name> {
             if (properties.isEmpty()) {
                 return emptyList()
             }
 
             // TODO: consolidate componentN() and copy(...) too?
 
-            // TODO: generate equals, hashCode, and toString only if needed
-            val equalsFunction = createSyntheticIrFunction(
-                Name.identifier("equals"),
-                components.irBuiltIns.booleanType,
-            ).apply {
-                valueParameters = listOf(
-                    createSyntheticIrParameter(this, Name.identifier("other"), components.irBuiltIns.anyNType)
-                )
+            val result = mutableListOf<Name>()
+
+            val superTypesContributedFunctionsThatMatchDataClassSyntheticMemberSignatures =
```

Comment:
There are *three* different cases here:

- `equals` in data class itself, then we should not generate synthetic
- non-final `equals` in super class, then we *should* generate synthetic (please check this case!)
- final `equals` in super class, then we again *should not* generate synthetic

Reply:
> There are _three_ different cases here:

Thank you for the clear categorization. This helps me figure out why my initial attempt looked so complicated. I've tried to collect contributed members in one go, but as you described below, ones in data class itself and ones in super class should be treated differently!

> * `equals` in data class itself, then we should not generate synthetic

Yeah, modality doesn't matter in this case. Note that unmuted tests in this PR whose file name is `alreadyDeclared.kt` cover this case.

> * non-final `equals` in super class, then we _should_ generate synthetic (please check this case!)

That's what I meant KT-6206, also covered here: https://github.com/JetBrains/kotlin/blob/master/compiler/testData/codegen/box/dataClasses/nonTrivialMemberInSuperClass.kt

because of this, I've added somewhat strange condition, but now, after separating contributed members in this type v.s. super types, the logic looks much clearer (to me, at least).

> * final `equals` in super class, then we again _should not_ generate synthetic

The unmuted test in this PR (`nonTrivialFinalMemberInSuperClass.kt`) covers this: https://github.com/JetBrains/kotlin/pull/3294/commits/6e5162916e9e6dc25aac728a0d8aa36336e1328c#diff-2bd0650f999b623ec18e3a6806e27516